servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/TransportObserver.java

```diff
@@ -24,6 +26,19 @@
      * Callback when transport starts initializing a new network connection.
      *
      * @return a new {@link ConnectionObserver} that provides visibility into events associated with a new connection
+     * @deprecated Use {@link #onNewConnection(Object, Object)}
      */
+    @Deprecated // FIXME: 0.42 - remove deprecated method
     ConnectionObserver onNewConnection();
+
+    /**
+     * Callback when transport starts initializing a new network connection.
+     *
+     * @param localAddress a local address of a new connection, if known
+     * @param remoteAddress a remote address of a new connection
+     * @return a new {@link ConnectionObserver} that provides visibility into events associated with a new connection
+     */
+    default ConnectionObserver onNewConnection(@Nullable Object localAddress, Object remoteAddress) {
```

Comment:
I tried to use `SocketAddress` as an upper bound for the client-side type, but it's too widespread across all API, making it impossible to be backward compatible or easy to migrate to. Went with an `Object` approach.

Reply:
Wouldn't the local address already be bound at this point? I'd expect at least `ANY_ADDR` with local port.

Reply:
It will be available on the server-side, but on the client-side it's `null` until we have `connectionEstablished`